### PR TITLE
wait time as a span

### PR DESCRIPTION
**Pull request by zhongzc**
Signed-off-by: zhongzc <zhongzc_arch@outlook.com>

Changes:
* One-dimensionalize spans. Remove `span_sets` in `TraceDetails`:
```diff
- pub span_sets: Vec<SpanSet>
+ pub spans: Vec<Span>,
+ pub properties: Properties,
```
* Introduce more relations naming `State` between spans: `State::Root`, `State::Settle`, `State::Local`, `State::Spawning`, `State::Scheduling`. `Root`, `Spawning` and `Scheduling` are spans which indicate waiting for execution.
* Treeize trace results in tests then more reasonable test cases can be written.

### CODE REVIEW

File: src/trace_local.rs [hunk 1]
```diff
@@ -42,14 +42,12 @@ static GLOBAL_ID_COUNTER: std::sync::atomic::AtomicU32 = std::sync::atomic::Atom
 
 #[inline]
 fn next_global_id_prefix() -> u32 {
-    GLOBAL_ID_COUNTER.fetch_add(1, std::sync::atomic::Ordering::SeqCst)
+    GLOBAL_ID_COUNTER.fetch_add(1, std::sync::atomic::Ordering::Relaxed)
```
Comment (Renkai) [inline src/trace_local.rs:45]: Ditto

File: src/trace_local.rs [hunk 2]
```diff
@@ -118,18 +139,18 @@ impl Drop for LocalTraceGuard {
     fn drop(&mut self) {
         let tl = unsafe { &mut *self.trace_local };
 
-        // fill the elapsed cycles of the first span
-        tl.spans[self.span_start_index].elapsed_cycles =
-            minstant::now().saturating_sub(tl.spans[self.span_start_index].begin_cycles);
+        // fill the elapsed cycles of the first span (except the leading span)
```
Comment (Renkai) [inline src/trace_local.rs:142]: Any explanation about why created a new concept `LeadingSpan` and change logic here?

File: src/collector.rs [hunk 3]
```diff
@@ -46,6 +88,6 @@ impl Drop for Collector {
     fn drop(&mut self) {
         self.inner
             .closed
-            .store(true, std::sync::atomic::Ordering::SeqCst);
+            .store(true, std::sync::atomic::Ordering::Relaxed);
```
[the remark this answers is not in the document]
Comment (zhongzc) [reply src/collector.rs:91]: Ordering is about two or more shared variables among multiple threads. If we don't have that synchronization semantics, relaxed is enough.

File: src/trace_local.rs (hunk 2 above)
Comment (zhongzc) [reply src/trace_local.rs:142]: A leading span is one of `Root`, `Spawning`, and `Scheduling`, which represents how a tracing context creates.

The elapsed time of a leading span means how long it waits from task creation to starting execution. 

A `State` span means the execution is starting.

File: src/lib.rs [hunk 4]
```diff
@@ -34,25 +34,87 @@ pub struct TraceDetails {
...
+/// ```text
+/// ------------------------------- TIME LINE -------------------------------------->
+///
+/// | <- enable trace
+/// +------------------------------------+-------------------------------+
+/// | state: Root, related_id: 0, id: 88 | state: Settle, related_id: 88 |
+/// +------------------------------------+-------------------------------+
+/// ```
+///
+/// ## Local
```
[the remark this answers is not in the document]
Comment (zhongzc) [reply src/lib.rs:67]: Respect to the enum name `State`.

File: src/collector.rs [hunk 5]
```diff
@@ -23,11 +23,53 @@ impl Collector {
 
     #[inline]
     pub fn collect(self) -> crate::TraceDetails {
+        let span_sets = self.collect_once();
```
[the remark this answers is not in the document]
Comment (zhongzc) [reply src/collector.rs:26]: Agree

File: src/trace_local.rs (hunk 2 above)
Comment (Renkai) [reply src/trace_local.rs:142]: > A leading span is one of `Root`, `Spawning`, and `Scheduling`, which represents how a tracing context creates.
> 
> The elapsed time of a leading span means how long it waits from creating the task to starting execution.
> 
> A `Settle` span means the execution is starting.

Maybe this response can add to the comments in L142

**Pull request merged**